### PR TITLE
Abdin branch

### PR DESCRIPTION
**Pull request by abdeen206**
testing these issues and getting support and feedback for the bug in profile updating 
#17 #18 #19 #22 #23 #24

### CODE REVIEW

File: db.sqlite3.sqbpro [hunk 1]
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="UTF-8"?><sqlb_project><db path="C:\Users\User_CL\Envs\RailsStockProject\db.sqlite3" foreign_keys="1" case_sensitive_like="0" temp_store="0" wal_autocheckpoint="1000" synchronous="2"/><attached/><window><current_tab id="1"/></window><tab_structure><column_width id="0" width="300"/><column_width id="1" width="0"/><column_width id="2" width="125"/><column_width id="3" width="3315"/><column_width id="4" width="0"/><expanded_item id="0" parent="1"/><expanded_item id="1" parent="1"/><expanded_item id="2" parent="1"/><expanded_item id="3" parent="1"/></tab_structure><tab_browse><current_table name="myapp_stock"/><default_encoding codec=""/><browse_table_settings><table schema="main" name="myapp_stock" show_row_id="0" encoding="" plot_x_axis="" unlock_view_pk=""><sort/><column_widths/><filter_values/><display_formats/><hidden_columns/><plot_y_axes/></table></browse_table_settings></tab_browse><tab_sql><sql name="SQL 1">DELETE FROM myapp_stock where symbol = symbol</sql><current_tab id="0"/></tab_sql></sqlb_project>
```
Comment (davidmeer20) [inline db.sqlite3.sqbpro:2]: What does the code/file do

File: myapp/forms.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#__author__ = Mohammad Abdin
+from django import forms
+from .models import Profile
+from django.contrib.auth.models import User
+
+
+
+""" this is the form for updating the user's profile."""
```
Comment (davidmeer20) [inline myapp/forms.py]: Docstrings are added after the class/function declaration

Comment (davidmeer20) [inline myapp/forms.py]: you can add to the docstring, called from http://127.0.0.1:8000/accounts/profile/edit page by clicking the update submit button

File: myapp/forms.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    first_name = forms.CharField(required=True)
+    last_name = forms.CharField(required=True)
+    job = forms.CharField(required=True)
+
+    class Meta:
+        model = Profile
+        fields = ('username', 'email', 'first_name', 'last_name', 'job' )
+
+    def clean_email(self):
+        username = self.cleaned_data.get('username')
```
Comment (davidmeer20) [inline myapp/forms.py]: docstring

File: myapp/forms.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        email = self.cleaned_data.get('email')
+
+        if email and User.objects.filter(email=email).exclude(username=username).count():
+            raise forms.ValidationError('This email address is already in use. Please supply a different email address.')
+        return email
+
+    def save(self, commit=True):
+        user = super(UpdateProfile, self).save(commit=False)
```
Comment (davidmeer20) [inline myapp/forms.py]: same, add docstirng

File: myapp/apps.py [hunk 5]
```diff
@@ -3,3 +3,6 @@
 
 class MyappConfig(AppConfig):
     name = 'myapp'
```
Comment (davidmeer20) [inline myapp/apps.py:5]: docstring

File: myapp/models.py [hunk 6]
```diff
@@ -1,13 +1,43 @@
...
 class Stock(models.Model):
-	symbol = models.CharField(max_length=12, primary_key=True)
-	name = models.CharField(max_length=64)
-	top_rank = models.IntegerField(null=True)
-	price = models.FloatField()
-	change = models.FloatField(null=True)
-	change_percent = models.FloatField()
-	market_cap = models.FloatField(null=True)
-	primary_exchange = models.CharField(null=True, max_length=32)
+    symbol = models.CharField(max_length=12, primary_key=True)
```
Comment (davidmeer20) [inline myapp/models.py]: why 12, check in google, bloomberg for the proper size

File: myapp/models.py [hunk 7]
```diff
@@ -1,13 +1,43 @@
...
+    name = models.CharField(max_length=64)
+    top_rank = models.IntegerField(null=True)
+    price = models.FloatField()
+    change = models.FloatField(null=True)
+    change_percent = models.FloatField()
+    market_cap = models.FloatField(null=True)
+    primary_exchange = models.CharField(null=True, max_length=32)
```
Comment (davidmeer20) [inline myapp/models.py]: max length should be smaller than 32

File: myapp/models.py [hunk 8]
```diff
@@ -1,13 +1,43 @@
...
+
+""" the user's profile model - created automatically by sending a signal when the user is created"""
```
Comment (davidmeer20) [inline myapp/models.py]: docstring bellow the class def.

File: myapp/models.py [hunk 9]
```diff
@@ -1,13 +1,43 @@
...
+class Profile(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    image = models.ImageField(default='default.jpg',upload_to='profile_pics')
+    # not working !!
```
Comment (davidmeer20) [inline myapp/models.py]: if not working, fix first, or comment out

File: myapp/models.py [hunk 10]
```diff
@@ -1,13 +1,43 @@
...
+    job = models.CharField(max_length=10, null=True, blank=True, default=None)
+    # TODO : add more info fields
+    STOCKS_CHOICES = ()
+    dbconn = sqlite3.connect('db.sqlite3')
+    cur = dbconn.cursor()
+
```
Comment (davidmeer20) [inline myapp/models.py]: remove blank line

File: myapp/models.py [hunk 11]
```diff
@@ -1,13 +1,43 @@
...
+    cur.execute("SELECT name,symbol FROM myapp_stock")
```
Comment (davidmeer20) [inline myapp/models.py]: add the sql statment as a const

File: myapp/views.py [hunk 12]
```diff
@@ -1,61 +1,87 @@
...
-		return render(request, 'register.html', {'page_title': 'Register'})
+    # If post -> register the user and redirect to main page
+    if request.method == 'POST':
+        username = request.POST.get('username')
+        firstname = request.POST.get('firstname')
+        lastname = request.POST.get('lastname')
+        email = request.POST.get('email')
+        password = request.POST.get('password')
+        #TODO: job = request.POST.get('job')
+
```
Comment (davidmeer20) [inline myapp/views.py]: remove blank line

File: myapp/views.py [hunk 13]
```diff
@@ -1,61 +1,87 @@
...
+""" for editing the user's profile(in db) and updating his info depending on the entered values"""
+@login_required
+def edit_profile(request):
+    args = {}
+    if request.method == 'POST':
+        form = UpdateProfile(request.POST, instance=request.user)
+        form.actual_user = request.user
+        if form.is_valid():
+            form.save()
+            return redirect('profile')
```
Comment (davidmeer20) [inline myapp/views.py]: is it working?

**Review by davidmeer20 — DISMISSED** [review]

File: db.sqlite3.sqbpro (hunk 1 above)
Comment (abdeen206) [reply db.sqlite3.sqbpro:2]: it's created automatically sometimes, when running the project. deleting it won;t make any change, so u can delete this file

**Comment by abdeen206** [comment]
profile fixed

File: myapp/forms.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+        if email and User.objects.filter(email=email).exclude(username=username).count():
+            raise forms.ValidationError('This email address is already in use. Please supply a different email address.')
+        return email
+
+    def save(self, commit=True):
+        """ this function saves the edits applied to the to user's profile"""
+        user = super(UpdateProfile, self).save(commit=False)
+        user.email = self.cleaned_data['email']
+        print("save profile .....")
```
Comment (farjun) [inline myapp/forms.py]: use logs, not prints inside a server. its much more reliable and fast

File: myapp/models.py [hunk 15]
```diff
@@ -1,13 +1,44 @@
...
+    primary_exchange = models.CharField(null=True, max_length=32)
+
+
+class Profile(models.Model):
+    """ the user's profile model - created automatically by sending a signal when the user is created"""
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    image = models.ImageField(default='default.jpg',upload_to='profile_pics')
+
+    job = models.CharField(max_length=20, null=True, default=None)
+    # TODO : add more info fields
```
Comment (farjun) [inline myapp/models.py]: remove TODO from your merges,

File: myapp/models.py [hunk 16]
```diff
@@ -1,13 +1,44 @@
...
+
+    dbconn = sqlite3.connect(DATABASE_NAME)
+    cur = dbconn.cursor()
+
+    cur.execute("SELECT name,symbol FROM {}".format(STOCKS_DATABASE))
+    rows = cur.fetchall()
+    stocks_choice = []
+    for row in rows:
```
Comment (farjun) [inline myapp/models.py]: replace with `stocks_choice.extend(rows)`
search for list's extend method on google to learn more

File: myapp/views.py [hunk 17]
```diff
@@ -1,61 +1,108 @@
...
+        profile_form = ProfileForm(instance=request.user.profile)
+        args = {}
+        # args.update(csrf(request))
+        args['form'] = form
+        args['profile_form'] = profile_form
+        return render(request, 'edit_profile.html', args)
+
+#
+# @login_required
+# def edit_profile(request):
```
Comment (farjun) [inline myapp/views.py]: remove commented code lines

File: .idea/vcs.xml [hunk 18]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (farjun) [inline .idea/vcs.xml]: remove all .idea files from your commit - they should not be added to master.
these are pycharm config files

**Review by farjun — DISMISSED** [review]
finish up with all the comments, click resolve and commit the fixed version.
looking good.

**Review by farjun — APPROVED** [review]

**Pull request merged**